genie-web/src/main/java/com/netflix/genie/web/services/impl/S3AttachmentService.java

```diff
@@ -0,0 +1,225 @@
+
+    private static final String METRICS_PREFIX = "genie.jobs.attachments.s3";
+    private static final String COUNT_DISTRIBUTION = METRICS_PREFIX + ".count.distribution";
+    private static final String LARGEST_SIZE_DISTRIBUTION = METRICS_PREFIX + ".largest.distribution";
+    private static final String TOTAL_SIZE_DISTRIBUTION = METRICS_PREFIX + ".totalSize.distribution";
+    private static final String SAVE_TIMER = METRICS_PREFIX + ".upload.timer";
+    private static final Set<URI> EMPTY_SET = ImmutableSet.of();
+    private static final String SLASH = "/";
+    private static final String S3 = "s3";
+    private final S3ClientFactory s3ClientFactory;
+    private final AttachmentServiceProperties properties;
+    private final MeterRegistry meterRegistry;
+    private final AmazonS3URI s3BaseURI;
+
+    /**
+     * Constructor.
+     *
+     * @param s3ClientFactory             the s3 client factory
+     * @param attachmentServiceProperties the service properties
+     * @param meterRegistry               the meter registry
+     */
+    public S3AttachmentService(
+        final S3ClientFactory s3ClientFactory,
+        final AttachmentServiceProperties attachmentServiceProperties,
+        final MeterRegistry meterRegistry
+    ) {
+        this.s3ClientFactory = s3ClientFactory;
+        this.properties = attachmentServiceProperties;
+        this.meterRegistry = meterRegistry;
+        this.s3BaseURI = new AmazonS3URI(attachmentServiceProperties.getLocationPrefix());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<URI> saveAttachments(
+        @Nullable final String jobId,
+        final Set<Resource> attachments
+    ) throws SaveAttachmentException {
+
+        // Track number of attachments, including zeroes
+        this.meterRegistry.summary(COUNT_DISTRIBUTION).record(attachments.size());
+
+        log.debug("Saving {} attachments for job request with id: {}", attachments.size(), jobId);
+
+        if (attachments.size() == 0) {
+            return EMPTY_SET;
+        }
+
+        // Check for attachment size limits
+        this.checkLimits(attachments);
+
+        final long start = System.nanoTime();
+        final Set<Tag> tags = Sets.newHashSet();
+        try {
+            // Upload all to S3
+            final Set<URI> attachmentURIs = this.uploadAllAttachments(jobId, attachments);
+            MetricsUtils.addSuccessTags(tags);
+            return attachmentURIs;
+        } catch (SaveAttachmentException e) {
+            log.error("Failed to save attachments (requested job id: {}): {}", jobId, e.getMessage(), e);
+            MetricsUtils.addFailureTagsWithException(tags, e);
+            throw e;
+        } finally {
+            this.meterRegistry
+                .timer(SAVE_TIMER, tags)
+                .record(System.nanoTime() - start, TimeUnit.NANOSECONDS);
+        }
+    }
+
+    private void checkLimits(final Set<Resource> attachments) throws SaveAttachmentException {
```

Comment:
Just pointing out that depending on how the lengths are calculated by a given `Resource` implementation this could result in 2 or 3 read throughs of all the bytes of attachments:
1. Spring at ingestion
2. This check
3. Saving to eventual destination

Reply:
Hmmm i assumed the multi-part upload has to provide size of parts so there's never any need to scan. Gonna double check.